### PR TITLE
Explain need for deprecated method

### LINKED ISSUE #22822: Explain need for deprecated DependencyHandlerScope constraints method
Method can't be removed, see notes in attached PR.

### PR DESCRIPTION
**Pull request by tresat**
Fixes #22822

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyHandlerScope.kt [hunk 1]
```diff
@@ -47,11 +45,6 @@ private constructor(
     override val delegate: DependencyHandler
         get() = dependencies
 
-    @Deprecated(replaceWith = ReplaceWith("constraints"), message = "This method shouldn't be called because the most specific variant should be preferred by the Kotlin compiler", level = DeprecationLevel.HIDDEN)
```
Comment (big-guy) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyHandlerScope.kt:50]: I think with `Deprecationlevel.HIDDEN` older plugins can still call this method, but new plugins can never access it.

That means we wouldn't have to remove this, but I'm not sure if that's worth it @eskatos wdyt?

Comment (tresat) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyHandlerScope.kt:50]: It was deprecated in Feb 2020, so it's been a while...

Comment (eskatos) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/DependencyHandlerScope.kt:50]: Yes, `Deprecationlevel.HIDDEN` works as you described Sterling.

This was introduced in https://github.com/gradle/gradle/commit/ac16c2e02fe4f2b311b4062e5187349b34fa7f3a as a workaround for wrong scoping.

I think it is fine to remove in a major :+1:

**Review by big-guy — APPROVED** [review]

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@eskatos, @big-guy - Without this method present (and deprecated as it currently is) the Kotlin DSL does not understand the `constraints` block.  This can be seen by running `DependencyConstraintsIntegrationTest#"dependency constraint is not included in resolution without a hard dependency"()` with the method present and deprecated, present but NOT deprecated, and absent.

Only in the first case is a constraint added properly, passing the test.  In the other 2 cases, the wrong receiver (the `DefaultDependencyHandler`) is used.  Because of this the `DependencyHandlerScope`'s `Configuration.invoke(dependencyNotation: Any)` extension method winds up being called, which results in a dependency, NOT a constraint, being added.

So it seems this can't be removed.  I'll update the message on the annotation to reflect this instead of removing, unless you guys have any ideas?

**Comment by eskatos** [comment]
Ah, looks like I forgot something about that workaround. Yes we need to keep it. This works around a flaw in the old `dependencies {}` block. :+1: to changing the message in the code

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**